wlroots/xwayland.py

```diff
@@ -0,0 +1,462 @@
+        ptr.no_touch_pointer_emulation = no_touch_pointer_emulation
+        return cls(ptr)
+
+    @property
+    def lazy(self) -> bool:
+        return self._ptr.lazy
+
+    @lazy.setter
+    def lazy(self, lazy: bool) -> None:
+        self._ptr.lazy = lazy
+
+    @property
+    def enable_wm(self) -> bool:
+        return self._ptr.enable_wm
+
+    @enable_wm.setter
+    def enable_wm(self, enable_wm: bool) -> None:
+        self._ptr.enable_wm = enable_wm
+
+    @property
+    def no_touch_pointer_emulation(self) -> bool:
+        return self._ptr.no_touch_pointer_emulation
+
+    @no_touch_pointer_emulation.setter
+    def no_touch_pointer_emulation(self, no_touch_pointer_emulation: bool) -> None:
+        self._ptr.no_touch_pointer_emulation = no_touch_pointer_emulation
+
+
+class Server(PtrHasData):
+    def __init__(self, ptr) -> None:
+        self._ptr = ptr
+
+        self.ready_event = Signal(ptr=ffi.addressof(self._ptr.events.ready))
+        self.destroy_event = Signal(ptr=ffi.addressof(self._ptr.events.destroy))
+
+    @classmethod
+    def create(cls, display: Display, options: ServerOptions) -> Server | None:
+        ptr = lib.wlr_xwayland_server_create(display._ptr, options._ptr)
+        if ptr == ffi.NULL:
+            return None
+        return cls(ptr)
+
+    def destroy(self) -> None:
+        lib.wlr_xwayland_server_destroy(self._ptr)
+
+
+class XWayland(PtrHasData):
+    def __init__(self, ptr) -> None:
+        self._ptr = ptr
+
+        self.ready_event = Signal(ptr=ffi.addressof(self._ptr.events.ready))
+        self.new_surface_event = Signal(
+            ptr=ffi.addressof(self._ptr.events.new_surface),
+            data_wrapper=Surface,
+        )
+        self.remove_startup_info_event = Signal(
+            ptr=ffi.addressof(self._ptr.events.remove_startup_info)
+        )
+
+    @classmethod
+    def create(
+        cls, display: Display, compositor: Compositor, lazy: bool
+    ) -> XWayland | None:
+        ptr = lib.wlr_xwayland_create(display._ptr, compositor._ptr, lazy)
+        if ptr == ffi.NULL:
+            return None
+        return cls(ptr)
+
+    @property
+    def display_name(self) -> str | None:
+        return str_or_none(self._ptr.display_name)
+
+    def destroy(self) -> None:
+        lib.wlr_xwayland_destroy(self._ptr)
+
+    def set_seat(self, seat: Seat) -> None:
+        lib.wlr_xwayland_set_seat(self._ptr, seat._ptr)
+
+    def set_cursor(
+        self,
+        pixels: list[int],
+        stride: int,
+        width: int,
+        height: int,
+        hotspot_x: int,
+        hotspot_y: int,
+    ) -> None:
+        lib.wlr_xwayland_set_cursor(
+            self._ptr, pixels, stride, width, height, hotspot_x, hotspot_y
+        )
+
+    def get_atom(self, name: str) -> int:
+        """Helper method to fetch an atom by name."""
+        conn = lib.xcb_connect(self._ptr.display_name, ffi.NULL)
+        if error := lib.xcb_connection_has_error(conn):
```

Comment:
Walrus operator! Cool!

I think we have this currently set to test on 3.7, which I think shouldn't work for this, so that tests pass is a but surprising. It is a bit surprising that this works. I wonder if we need to get flake8 running on the oldest version of Python that is supported, rather than the latest.

Reply:
Or perhaps both? Anything in between will almost certainly be valid.